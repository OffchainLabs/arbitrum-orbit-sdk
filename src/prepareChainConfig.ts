import { ChainConfig, ChainConfigArbitrumParams } from './types/ChainConfig';

export const defaults = {
  homesteadBlock: 0,
  daoForkBlock: null,
  daoForkSupport: true,
  eip150Block: 0,
  eip150Hash: '0x0000000000000000000000000000000000000000000000000000000000000000',
  eip155Block: 0,
  eip158Block: 0,
  byzantiumBlock: 0,
  constantinopleBlock: 0,
  petersburgBlock: 0,
  istanbulBlock: 0,
  muirGlacierBlock: 0,
  berlinBlock: 0,
  londonBlock: 0,
  clique: {
    period: 0,
    epoch: 0,
  },
  arbitrum: {
    EnableArbOS: true,
    AllowDebugPrecompiles: false,
    DataAvailabilityCommittee: false,
    InitialArbOSVersion: 20,
    GenesisBlockNum: 0,
    MaxCodeSize: 24_576,
    MaxInitCodeSize: 49_152,
  },
};

export type PrepareChainConfigParams = Pick<ChainConfig, 'chainId'> &
  Partial<Omit<ChainConfig, 'chainId' | 'arbitrum'>> & {
    arbitrum: Pick<ChainConfigArbitrumParams, 'InitialChainOwner'> &
      Partial<Omit<ChainConfigArbitrumParams, 'InitialChainOwner'>>;
  };

export function prepareChainConfig(params: PrepareChainConfigParams): ChainConfig {
  return {
    ...defaults,
    ...params,
    clique: { ...defaults.clique, ...params.clique },
    arbitrum: { ...defaults.arbitrum, ...params.arbitrum },
  };
}
