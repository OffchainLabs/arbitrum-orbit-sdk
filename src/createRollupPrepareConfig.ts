import { ChainConfig } from './types/ChainConfig';
import { CreateRollupFunctionInputs } from './types/createRollupTypes';
import { prepareChainConfig } from './prepareChainConfig';
import { defaults as defaultsCommon } from './createRollupPrepareDeploymentParamsConfigDefaults';
// importing for jsdoc @link to work
import { createRollupPrepareDeploymentParamsConfig } from './createRollupPrepareDeploymentParamsConfig';
import { Prettify } from './types/utils';

export type CreateRollupPrepareConfigResult = CreateRollupFunctionInputs[0]['config'];

type RequiredKeys = 'chainId' | 'owner';
type RequiredParams = Pick<CreateRollupPrepareConfigResult, RequiredKeys>;
type OptionalParams = Partial<Omit<CreateRollupPrepareConfigResult, 'chainConfig' | RequiredKeys>>;

export type CreateRollupPrepareConfigParams = Prettify<
  RequiredParams & { chainConfig?: ChainConfig } & OptionalParams
>;

export const defaults = {
  ...defaultsCommon,
  confirmPeriodBlocks: BigInt(150),
  sequencerInboxMaxTimeVariation: {
    delayBlocks: BigInt(5_760),
    futureBlocks: BigInt(48),
    delaySeconds: BigInt(86_400),
    futureSeconds: BigInt(3_600),
  },
} as const;

/**
 * @deprecated Will be removed in a future release. Please use {@link createRollupPrepareDeploymentParamsConfig} instead.
 */
export function createRollupPrepareConfig({
  chainConfig,
  ...params
}: CreateRollupPrepareConfigParams): CreateRollupPrepareConfigResult {
  return {
    ...defaults,
    ...params,
    chainConfig: JSON.stringify(
      chainConfig ??
        prepareChainConfig({
          chainId: Number(params.chainId),
          arbitrum: { InitialChainOwner: params.owner },
        }),
    ),
  };
}
