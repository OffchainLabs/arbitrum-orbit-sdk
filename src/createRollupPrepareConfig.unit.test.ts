import { expect, it } from 'vitest';

import { prepareChainConfig } from './prepareChainConfig';
import { createRollupPrepareConfig } from './createRollupPrepareConfig';

const chainId = 69_420n;
const vitalik: `0x${string}` = '0xd8da6bf26964af9d7eed9e03e53415d37aa96045';

it('creates config with defaults', () => {
  expect(
    createRollupPrepareConfig({
      owner: vitalik,
      chainId,
    }),
  ).toMatchSnapshot();
});

it('creates config with overrides', () => {
  expect(
    createRollupPrepareConfig({
      owner: vitalik,
      chainId,
      chainConfig: prepareChainConfig({
        chainId: 69_420,
        arbitrum: {
          InitialChainOwner: vitalik,
          InitialArbOSVersion: 30,
          DataAvailabilityCommittee: true,
        },
      }),
      confirmPeriodBlocks: 4200n,
      extraChallengeTimeBlocks: 5n,
      loserStakeEscrow: '0x0000000000000000000000000000000000000001',
      sequencerInboxMaxTimeVariation: {
        delayBlocks: 200n,
        delaySeconds: 5n,
        futureBlocks: 100n,
        futureSeconds: 1n,
      },
      stakeToken: '0x0000000000000000000000000000000000000002',
      wasmModuleRoot: '0xWasmModuleRoot',
    }),
  ).toMatchSnapshot();
});
