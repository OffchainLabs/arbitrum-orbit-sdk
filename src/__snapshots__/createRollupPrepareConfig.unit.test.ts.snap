// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`creates config with defaults 1`] = `
{
  "baseStake": 100000000000000000n,
  "chainConfig": "{\\"homesteadBlock\\":0,\\"daoForkBlock\\":null,\\"daoForkSupport\\":true,\\"eip150Block\\":0,\\"eip150Hash\\":\\"0x0000000000000000000000000000000000000000000000000000000000000000\\",\\"eip155Block\\":0,\\"eip158Block\\":0,\\"byzantiumBlock\\":0,\\"constantinopleBlock\\":0,\\"petersburgBlock\\":0,\\"istanbulBlock\\":0,\\"muirGlacierBlock\\":0,\\"berlinBlock\\":0,\\"londonBlock\\":0,\\"clique\\":{\\"period\\":0,\\"epoch\\":0},\\"arbitrum\\":{\\"EnableArbOS\\":true,\\"AllowDebugPrecompiles\\":false,\\"DataAvailabilityCommittee\\":false,\\"InitialArbOSVersion\\":20,\\"GenesisBlockNum\\":0,\\"MaxCodeSize\\":24576,\\"MaxInitCodeSize\\":49152,\\"InitialChainOwner\\":\\"0xd8da6bf26964af9d7eed9e03e53415d37aa96045\\"},\\"chainId\\":69420}",
  "chainId": 69420n,
  "confirmPeriodBlocks": 150n,
  "extraChallengeTimeBlocks": 0n,
  "genesisBlockNum": 0n,
  "loserStakeEscrow": "0x0000000000000000000000000000000000000000",
  "owner": "0xd8da6bf26964af9d7eed9e03e53415d37aa96045",
  "sequencerInboxMaxTimeVariation": {
    "delayBlocks": 5760n,
    "delaySeconds": 86400n,
    "futureBlocks": 48n,
    "futureSeconds": 3600n,
  },
  "stakeToken": "0x0000000000000000000000000000000000000000",
  "wasmModuleRoot": "0x8b104a2e80ac6165dc58b9048de12f301d70b02a0ab51396c22b4b4b802a16a4",
}
`;

exports[`creates config with overrides 1`] = `
{
  "baseStake": 100000000000000000n,
  "chainConfig": "{\\"homesteadBlock\\":0,\\"daoForkBlock\\":null,\\"daoForkSupport\\":true,\\"eip150Block\\":0,\\"eip150Hash\\":\\"0x0000000000000000000000000000000000000000000000000000000000000000\\",\\"eip155Block\\":0,\\"eip158Block\\":0,\\"byzantiumBlock\\":0,\\"constantinopleBlock\\":0,\\"petersburgBlock\\":0,\\"istanbulBlock\\":0,\\"muirGlacierBlock\\":0,\\"berlinBlock\\":0,\\"londonBlock\\":0,\\"clique\\":{\\"period\\":0,\\"epoch\\":0},\\"arbitrum\\":{\\"EnableArbOS\\":true,\\"AllowDebugPrecompiles\\":false,\\"DataAvailabilityCommittee\\":true,\\"InitialArbOSVersion\\":30,\\"GenesisBlockNum\\":0,\\"MaxCodeSize\\":24576,\\"MaxInitCodeSize\\":49152,\\"InitialChainOwner\\":\\"0xd8da6bf26964af9d7eed9e03e53415d37aa96045\\"},\\"chainId\\":69420}",
  "chainId": 69420n,
  "confirmPeriodBlocks": 4200n,
  "extraChallengeTimeBlocks": 5n,
  "genesisBlockNum": 0n,
  "loserStakeEscrow": "0x0000000000000000000000000000000000000001",
  "owner": "0xd8da6bf26964af9d7eed9e03e53415d37aa96045",
  "sequencerInboxMaxTimeVariation": {
    "delayBlocks": 200n,
    "delaySeconds": 5n,
    "futureBlocks": 100n,
    "futureSeconds": 1n,
  },
  "stakeToken": "0x0000000000000000000000000000000000000002",
  "wasmModuleRoot": "0xWasmModuleRoot",
}
`;
